src/Riok.Mapperly/Descriptors/MappingBuilders/SpanMappingBuilder.cs

```diff
@@ -0,0 +1,270 @@
+            );
+        }
+
+        if (
+            !ctx.ObjectFactories.TryFindObjectFactory(ctx.Source, ctx.Target, out var objectFactory)
+            && !ctx.Target.HasAccessibleParameterlessConstructor()
+        )
+        {
+            return MapSpanArrayToEnumerableMethod(ctx);
+        }
+
+        if (ctx.Target.ImplementsGeneric(ctx.Types.Get(typeof(Stack<>)), out _))
+            return CreateForEach(nameof(Stack<object>.Push), objectFactory);
+
+        if (ctx.Target.ImplementsGeneric(ctx.Types.Get(typeof(Queue<>)), out _))
+            return CreateForEach(nameof(Queue<object>.Enqueue), objectFactory);
+
+        // create a foreach loop with add calls if source is not an array
+        // and ICollection.Add(T): void is implemented and not explicit
+        // ensures add is not called and immutable types
+        if (!ctx.Target.IsArrayType() && ctx.Target.HasImplicitGenericImplementation(ctx.Types.Get(typeof(ICollection<>)), AddMethodName))
+            return CreateForEach(AddMethodName, objectFactory);
+
+        return MapSpanArrayToEnumerableMethod(ctx);
+
+        ExistingTargetMappingMethodWrapper CreateForEach(string methodName, ObjectFactory? factory)
+        {
+            var ensureCapacityStatement = EnsureCapacityBuilder.TryBuildEnsureCapacity(ctx.Source, ctx.Target, ctx.Types);
+            return new ForEachAddEnumerableMapping(ctx.Source, ctx.Target, elementMapping, factory, methodName, ensureCapacityStatement);
+        }
+    }
+
+    private static TypeMapping BuildToArrayOrMap(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        if (!elementMapping.IsSynthetic)
+            return new ArrayForMapping(ctx.Source, ctx.Target, elementMapping, elementMapping.TargetType);
+
+        return new SourceObjectMethodDelegateMapping(ctx.Source, ctx.Target, ToArrayMethodName);
+    }
+
+    private static TypeMapping? BuildEnumerableToSpan(MappingBuilderContext ctx, ITypeMapping elementMapping)
+    {
+        var typedArray = ctx.Types.GetArrayType(elementMapping.TargetType);
+        if (ctx.FindOrBuildMapping(ctx.Source, typedArray.NonNullable()) is not { } arrayMapping)
+            return null;
+
+        return new CastMapping(ctx.Source, ctx.Target, arrayMapping);
+    }
+
+    private static TypeMapping? MapSpanArrayToEnumerableMethod(MappingBuilderContext ctx)
+    {
+        var enumeratedType = ((INamedTypeSymbol)ctx.Source).TypeArguments[0];
+        var typedArray = ctx.Types.GetArrayType(enumeratedType);
+        if (ctx.FindOrBuildMapping(typedArray.NonNullable(), ctx.Target) is not { } arrayMapping)
+            return null;
+
+        return new SourceObjectMethodDelegateMapping(ctx.Source, ctx.Target, ToArrayMethodName, arrayMapping);
+    }
+
+    private static bool IsTargetImmutableCollection(MappingBuilderContext ctx)
+    {
+        if (SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableArray<>))))
```

Comment:
Similar checks in the `EnumerableMappingBuilder` and `MemoryMappingBuilder`. I think it would make sense to extract the enum `CollectionType` and create a method in `WellKnownTypes` `CollectionType GetCollectionType(ITypeSymbol type)`. `CollectionType` could include all collection types handled by Mapperly (eg. `List, ImmutableList, Array, Span, ReadOnlySpan, Memory, ...`). The enum could then have extension methods (eg. `bool IsSpan()`, `bool IsImmutable()`, ...).